lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp

```diff
@@ -0,0 +1,292 @@
+  inlineAnnoClass = StringAttr::get(cx, "firrtl.transforms.BlackBoxInlineAnno");
+  pathAnnoClass = StringAttr::get(cx, "firrtl.transforms.BlackBoxPathAnno");
+  resourceAnnoClass =
+      StringAttr::get(cx, "firrtl.transforms.BlackBoxResourceAnno");
+
+  // Determine the target directory and resource file name from the
+  // annotations present on the circuit operation.
+  targetDir = ".";
+  resourceFileName = "firrtl_black_box_resource_files.f";
+
+  if (auto attrs = circuitOp->getAttrOfType<ArrayAttr>("annotations")) {
+    for (auto attr : attrs) {
+      auto anno = attr.dyn_cast<DictionaryAttr>();
+      if (!anno)
+        continue;
+      auto cls = anno.getAs<StringAttr>("class");
+
+      // Handle target dir annotation.
+      if (cls == targetDirAnnoClass) {
+        auto x = anno.getAs<StringAttr>("targetDir");
+        if (!x) {
+          circuitOp->emitError(targetDirAnnoClass.getValue() +
+                               " annotation missing \"targetDir\" attribute");
+          signalPassFailure();
+          continue;
+        }
+        targetDir = x.getValue();
+        continue;
+      }
+
+      // Handle resource file name annotation.
+      if (cls == resourceFileNameAnnoClass) {
+        auto x = anno.getAs<StringAttr>("resourceFileName");
+        if (!x) {
+          circuitOp->emitError(
+              resourceFileNameAnnoClass.getValue() +
+              " annotation missing \"resourceFileName\" attribute");
+          signalPassFailure();
+          continue;
+        }
+        resourceFileName = x.getValue();
+        continue;
+      }
+    }
+  }
+
+  LLVM_DEBUG(llvm::dbgs() << "Black box target directory: " << targetDir << "\n"
+                          << "Black box resource file name: "
+                          << resourceFileName << "\n");
+
+  // Gather the relevant annotations on all modules in the circuit.
+  OpBuilder builder(&getContext());
+  builder.setInsertionPointToEnd(getOperation()->getBlock());
+
+  for (auto &op : *circuitOp.getBody()) {
+    if (!isa<FModuleOp>(op) && !isa<FExtModuleOp>(op))
+      continue;
+    if (auto attrs = op.getAttrOfType<ArrayAttr>("annotations")) {
+      for (auto attr : attrs) {
+        if (auto anno = attr.dyn_cast<DictionaryAttr>())
+          runOnAnnotation(&op, anno, builder);
+      }
+    }
+  }
+
+  // If we have emitted any files, generate a file list operation that
+  // documents the additional annotation-controlled file listing to be
+  // created.
+  if (!emittedFiles.empty()) {
+    auto trueAttr = BoolAttr::get(cx, true);
+    std::string output;
+    llvm::raw_string_ostream os(output);
+    for (auto &file : emittedFiles) {
+      const auto &fileName = file.fileName.getValue();
+      // Exclude Verilog header files since we expect them to be included
+      // explicitly by compiler directives in other source files.
+      auto ext = llvm::sys::path::extension(fileName);
+      bool exclude = (ext == ".h" || ext == ".vh" || ext == ".svh");
```

Comment:
I didn't realize this was something the Blackbox transform would do. Nice work staying true to the Scala code. :+1: